src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -0,0 +1,391 @@
+ *      key should only exists once in the WHERE clause. So basically,
+ *          SELECT ... FROM dist_table WHERE dist_key = X
+ *   - No returning for UPDATE/DELETE queries
+ */
+bool
+FastPathRouterQuery(Query *query)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	List *rangeTableList = NIL;
+	FromExpr *joinTree = query->jointree;
+	Node *quals = NULL;
+	Oid distributedTableId = InvalidOid;
+	Var *distributionKey = NULL;
+	DistTableCacheEntry *cacheEntry = NULL;
+	List *varClauseList = NIL;
+	ListCell *varClauseCell = NULL;
+	int partitionColumnReferenceCount = 0;
+
+	if (!EnableFastPathRouterPlanner)
+	{
+		return false;
+	}
+
+	if (!(query->commandType == CMD_SELECT || query->commandType == CMD_UPDATE ||
+		  query->commandType == CMD_DELETE))
+	{
+		return false;
+	}
+
+	/*
+	 * We want to deal with only very simple select queries. Some of the
+	 * checks might be too restrictive, still we prefer this way.
+	 */
+	if (query->cteList != NIL || query->returningList != NIL ||
+		query->hasSubLinks || query->setOperations != NULL ||
+		query->hasForUpdate || query->hasTargetSRFs ||
+		query->hasModifyingCTE)
+	{
+		return false;
+	}
+
+	/* make sure that the only range table in FROM clause */
+	if (list_length(query->rtable) != 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = (RangeTblEntry *) linitial(query->rtable);
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+
+	/*
+	 * We don't want to deal with potentially overlapping
+	 * append/range distributed tables.
+	 */
+	distributedTableId = rangeTableEntry->relid;
+	cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH ||
+		  cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		return false;
+	}
+
+	/* WHERE clause should not be empty for distributed tables */
+	if (joinTree == NULL ||
+		(cacheEntry->partitionMethod != DISTRIBUTE_BY_NONE && joinTree->quals == NULL))
+	{
+		return false;
+	}
+
+	/*
+	 * Pull all range table entries. We prefer this to make sure that
+	 * there is no subqueries in the any part of the query, including
+	 * the from clause, where clause, target list or HAVING clause etc.
+	 */
+	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+	if (list_length(rangeTableList) != 1)
+	{
+		return false;
+	}
+
+	/* convert list of expressions into expression tree */
+	quals = joinTree->quals;
+	if (quals != NULL && IsA(quals, List))
+	{
+		quals = (Node *) make_ands_explicit((List *) quals);
+	}
+
+	/*
+	 * Distribution column must be used in a simple equality match check and it must be
+	 * place at top level conjustion operator. In simple words, we should have
+	 *	    WHERE dist_key = VALUE [AND  ....];
+	 *
+	 * We also skip this check for reference tables.
+	 */
+	distributionKey = PartitionColumn(distributedTableId, 1);
+	if (distributionKey != NULL &&
+		!ColumnMatchExpressionAtTopLevelConjunction(quals, distributionKey))
```

Comment:
Wouldn't it be simpler to iterate over a list of implict ANDs and find a `<distribution column> = <Const>` / `<Const> = <distribution column>` clause?

Reply:
We've talked this in person. Noting for readers: `make_ands_implicit()` doesn't recurse into the expression tree. So, if we have `a AND (b and (c and d))` we wouldn't get `a,b,d,c`. Thus, we need to recurse anyway. So, that approach would also have a similar implementation.